### PR TITLE
Support right precedence hack-style pipeline

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Hack-style pipeline operators are right-associative. `babel/parser@7.15.5` treats it as left-associative, `babel/parser@7.15.6` treats it as right-associative.

ref: https://github.com/babel/babel/pull/13668
ref: https://github.com/babel/babel/pull/13668#issuecomment-896965830

Clsoes #11540

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -240,14 +240,14 @@ async function * f () {
...
-  |> (await bar.bar(%))
-  |> console.log(%);
+|> (await %)
+|> % || throw new Error(\`foo \${bar1}\`)
+|> bar2(%, ", ")
+|> bar3(%)
+|> % + "!"
+|> new Bar.Foo(%)
+|> (await bar.bar(%))
+|> console.log(%);
```
Comment (sosukesuzuki) [inline tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap:250]: I'm not sure the best way, but following [official explainer](https://github.com/tc39/proposal-pipeline-operator#real-world-examples), we don't add indent on here.

**Comment by fisker** [comment]
Let me understand the problem here

1. We can't know the ast is a hack-style pipeline, even if we search for `TopicReference`, because `TopicReference` is not required in pipeline. So we have to store the `plugins`, right?
2. This ast doesn't print right because we are trying to flat the `BinaryExpression`. What if we don't try to flat `|>`? (To be honest: I don't like prettier flat my operators, even simple `+`) We already have [`**`](https://github.com/prettier/prettier/blob/41d77d6c13b27631726c7f4ef5cd56978c8c6da8/src/language-js/utils.js#L1050), will the output ugly?

**Comment by sosukesuzuki** [comment]
@fisker

> We can't know the ast is a hack-style pipeline, even if we search for TopicReference, because TopicReference is not required in pipeline. So we have to store the plugins, right?

No! `TopicReference` is required in hack pipeline. So we can detect it in the way you said. I had no such idea.

> This ast doesn't print right because we are trying to flat the BinaryExpression.

Why? Hack pipeline is right-associative operators, so it is no problem to flat it?

> What if we don't try to flat |>? (To be honest: I don't like prettier flat my operators, even simple logical operators) We already have ** , will the output ugly?

When we don't flat it...

input:

```js
const searchResults$ = fromEvent(document.querySelector('input'), 'input')
  |> map(%, event => event.target.value)
  |> filter(%, searchText => searchText.length > 2)
  |> debounce(%, 300)
  |> distinctUntilChanged(%)
  |> switchMap(%, searchText => queryApi(searchText) |> retry(%, 3))
  |> share(%);
```

output:

```js
const searchResults$ =
  fromEvent(document.querySelector("input"), "input")
  |> (map(%, (event) => event.target.value)
    |> (filter(%, (searchText) => searchText.length > 2)
      |> (debounce(%, 300)
        |> (distinctUntilChanged(%)
          |> (switchMap(%, (searchText) => queryApi(searchText) |> retry(%, 3))
            |> share(%))))));
```

**Comment by fisker** [comment]
> This ast doesn't print right because we are trying to flat the BinaryExpression.
> Why? Hack pipeline is right-associative operators, so it is no problem to flat it?

I don't know, my guess.

> When we don't flat it...

That's ugly, but maybe not caused by flat? If we can know it's right-associative, why not just print differently, like not add parens and intentions?

File: tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (fisker) [reply tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap:250]: It's clear for me to indent.

Comment (sosukesuzuki) [reply tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap:250]: I prefer no indentation. Also other languages code formatter outputs following:

## Elixir (by `mix format`)

**Input:**

```ex
100 |> foo(1) |> bar(2) |> baz(3) |> bar(2) |> baz(3) |> bar(2) |> baz(3)
```

**output:**

```ex
100
|> foo(1)
|> bar(2)
|> baz(3)
|> bar(2)
|> baz(3)
|> bar(2)
|> baz(3)
```

## F# (by [Fantomas](https://github.com/fsprojects/fantomas))

**input:**

```fs
100 |> function1 |> function2 |> function1 |> function2 |> function1 |> function2 |> function1 |> function2 
```

**output:**

```fs
100
|> function1
|> function2
|> function1
|> function2
|> function1
|> function2
|> function1
|> function2
```

File: src/language-js/parse/babel.js [hunk 2]
```diff
@@ -149,7 +149,8 @@ function createParse(parseMethod, ...optionsCombinations) {
       throw createBabelParseError(error);
     }
 
-    return postprocess(ast, { ...opts, originalText: text });
```
Comment (sosukesuzuki) [inline src/language-js/parse/babel.js:152]: out of scope: `opts` is big object, so maybe it is good that we don't copy it for other parsers too.

**Comment by sosukesuzuki** [comment]
@fisker Can you review again?

File: tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (fisker) [reply tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap:250]: This can't convince me. I'd like to hear opinions from other maintainers.

**Comment by fisker** [comment]
`  1 + 100 |> a(%)` means `a(1 + 100)`, not `1 + a(100)`?

**Comment by sosukesuzuki** [comment]
> 1 + 100 |> a(%) means a(1 + 100)

yes.

([Babel](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=IwAg1CwAxSA-A-EBDAFAUgJRA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Cstage-1&prettier=false&targets=&version=7.15.7&externalPlugins=&assumptions=%7B%7D))

**Input:**

```js
1 + 100 |> a(%)
```

**Output:**

```js
"use strict";

var _ref;

_ref = 1 + 100, a(_ref);
```

File: tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (sosukesuzuki) [reply tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap:250]: Okay, I have no strong opinion for this. So for now, let's use indentation version, same as current Prettier.

File: tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -257,34 +257,34 @@ function createPerson(attrs) {
...
-    |> a.b?.(%)
-    |> a.b?.(%)
+      |> (yield %)
+      |> (await %)
+      |> y(%)
+      |> a.b(%)
+      |> a.b(%)
+      |> a.b(%)
+      |> a.b?.(%)
+      |> a.b?.(%)
```
Comment (sosukesuzuki) [inline tests/format/js/pipeline-operator/__snapshots__/jsfmt.spec.js.snap:287]: @fisker I've updated snapshots with indentation version (https://github.com/prettier/prettier/pull/11557#discussion_r718039101). What do you  think this change?

**Comment by sosukesuzuki** [comment]
I don't like that this PR is blocker to update babel/parser. So I'll merge this with the unindent version for now. Let's discuss in another issue. Since this syntax is not used by most users, there is no pain to change the behavior.

**Pull request merged**